libfwupdplugin/fu-msgpack.c

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2024 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1-or-later
+ */
+
+#define G_LOG_DOMAIN "FuMsgpack"
+
+#include "config.h"
+
+#include "fu-msgpack-item-private.h"
+#include "fu-msgpack.h"
+
+/**
+ * fu_msgpack_new:
+ * @buf: data blob
+ * @error: (nullable): optional return location for an error
+ *
+ * Parses a buffer into messagepack items.
+ *
+ * Returns: (transfer container) (element-type FuMsgpackItem): items, or %NULL on error
+ *
+ * Since: 2.0.0
+ **/
+GPtrArray *
+fu_msgpack_parse(GByteArray *buf, GError **error)
+{
+	gsize offset = 0;
+	g_autoptr(GPtrArray) items = g_ptr_array_new_with_free_func((GDestroyNotify)g_object_unref);
+
+	g_return_val_if_fail(buf != NULL, NULL);
+	g_return_val_if_fail(error == NULL || *error == NULL, NULL);
+
+	while (offset < buf->len) {
+		g_autoptr(FuMsgpackItem) item = NULL;
+		item = fu_msgpack_item_parse(buf, &offset, error);
+		if (item == NULL)
+			return NULL;
+		g_ptr_array_add(items, g_steal_pointer(&item));
+	}
+
+	return g_steal_pointer(&items);
+}
+
+/**
+ * fu_msgpack_write:
+ * @items: (element-type FuMsgpackItem): items
+ * @error: (nullable): optional return location for an error
+ *
+ * Writes messagepack items into a buffer.
+ *
+ * Returns: (transfer container): buffer, or %NULL on error
+ *
+ * Since: 2.0.0
+ **/
+GByteArray *
+fu_msgpack_write(GPtrArray *items, GError **error)
+{
+	g_autoptr(GByteArray) buf = g_byte_array_new();
+
+	g_return_val_if_fail(items != NULL, NULL);
+	g_return_val_if_fail(error == NULL || *error == NULL, NULL);
```

Comment:
why can't a user call this with a NULL error?

Reply:
It's the same mental gymnastics as the other callers, essentially what it's saying is that a NULL error is okay, and a GError ** pointing to a NULL GError * is okay, but what's not okay is if the *error is non-NULL -- which is practice means "return with NULL if the GError is already set" which stops us getting the "The GError was overwritten with a new value" critical error on the new error path -- i.e. failing fast if we forget to return from a previous error with failure.